Xamarin.Forms.Xaml/MarkupExtensions/OnIdiomExtension.cs

```diff
@@ -0,0 +1,78 @@
+﻿using System;
+using System.Globalization;
+using System.Reflection;
+
+namespace Xamarin.Forms.Xaml
+{
+	[ContentProperty("Default")]
+	public class OnIdiomExtension : IMarkupExtension
+	{
+		// See Device.Idiom
+
+		public object Default { get; set; }
+		public object Unsupported { get; set; }
+		public object Phone { get; set; }
+		public object Tablet { get; set; }
+		public object Desktop { get; set; }
+		public object TV { get; set; }
+		public object Watch { get; set; }
+
+		public IValueConverter Converter { get; set; }
+
+		public object ConverterParameter { get; set; }
+
+		public object ProvideValue(IServiceProvider serviceProvider)
+		{
+			var lineInfo = (serviceProvider.GetService(typeof(IXmlLineInfoProvider)) as IXmlLineInfoProvider)?.XmlLineInfo;
+			if (Default == null && Unsupported == null && Phone == null &&
+				Tablet == null && Desktop == null && TV == null && Watch == null)
+			{
+				throw new XamlParseException("OnIdiomExtension requires a value to be specified for at least one idiom or Default.", lineInfo ?? new XmlLineInfo());
+			}
+
+			var valueProvider = serviceProvider.GetService<IProvideValueTarget>() ?? throw new ArgumentException();
```

Comment:
`serviceProvider?.`

Reply:
Can the service provider itself be null?